truffle/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/TruffleInstrument.java

```diff
@@ -0,0 +1,299 @@
+         * @throws IllegalStateException if the method is called later than from
+         *             {@link #onCreate(com.oracle.truffle.api.instrumentation.TruffleInstrument.Env) }
+         *             method
+         */
+        public void registerService(Object service) {
+            if (services == null) {
+                throw new IllegalStateException();
+            }
+            services.add(service);
+        }
+
+        Object[] onCreate(TruffleInstrument instrumentation) {
+            List<Object> arr = new ArrayList<>();
+            services = arr;
+            try {
+                instrumentation.onCreate(this);
+            } finally {
+                services = null;
+            }
+            return arr.toArray();
+        }
+
+        /**
+         * Evaluates source of (potentially different) language using the current context.The names
+         * of arguments are parameters for the resulting {#link CallTarget} that allow the
+         * <code>source</code> to reference the actual parameters passed to
+         * {@link CallTarget#call(java.lang.Object...)}.
+         *
+         * @param source the source to evaluate
+         * @param argumentNames the names of {@link CallTarget#call(java.lang.Object...)} arguments
+         *            that can be referenced from the source
+         * @return the call target representing the parsed result
+         * @throws IOException if the parsing or evaluation fails for some reason
+         */
+        @SuppressWarnings("static-method")
+        public CallTarget parse(Source source, String... argumentNames) throws IOException {
+            return InstrumentationHandler.ACCESSOR.parse(null, source, null, argumentNames);
+        }
+
+    }
+
+    /**
+     * Annotation to register an {@link TruffleInstrument instrument} implementations for automatic
+     * discovery.
+     */
+    @Retention(RetentionPolicy.SOURCE)
+    @Target(ElementType.TYPE)
+    public @interface Registration {
+
+        /**
+         * A custom machine identifier for this instrumentation. If not defined then the fully
+         * qualified class name is used.
+         */
+        String id() default "";
+
+        /**
+         * The name of the instrumentation in an arbitrary format for humans.
+         */
+        String name() default "";
+
+        /**
+         * The version for instrumentation in an arbitrary format.
+         */
+        String version() default "";
+
+        /**
+         * Whether or not the annotated instrumentation should be enabled by default. If autostart
+         * is <code>true</code> then {@link TruffleInstrument#onCreate(Env)} is invoked for this
+         * instrument just before the first guest language code is executed. If autostart is
+         * <code>false</code> then the instrument needs to be enabled using the engine.
+         */
+        boolean autostart() default false;
+
```

Comment:
I am not anymore convinced that *autostart*=**true** is needed.

Debugger certainly doesn't need it and even in case of profiler I don't see any reason for profiling being always on for all executions. It may be "always on" in some configurations, but that isn't something to be decided by the instrument, but rather by somebody who configures the **PolyglotEngine**.

There is *static Debugger.find*(PolyglotEngine) that needs to be called by V8 SocketServer or by JPDA connector to turn the debugging on. Something similar, imho, can be done by other instruments as well.